subsys/net/openthread/rpc/client/ot_rpc_dns_client.c

```diff
@@ -0,0 +1,551 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <openthread/dns_client.h>
+
+#include <nrf_rpc/nrf_rpc_serialize.h>
+#include <nrf_rpc_cbor.h>
+#include <ot_rpc_common.h>
+#include <ot_rpc_ids.h>
+#include <ot_rpc_common.h>
+
+#include <zcbor_encode.h>
+
+static otDnsQueryConfig default_config;
+
+static otError decode_ot_error(struct nrf_rpc_cbor_ctx *ctx)
+{
+	otError error;
+
+	if (!zcbor_uint_decode(ctx->zs, &error, sizeof(error))) {
+		error = OT_ERROR_PARSE;
+	}
+
+	nrf_rpc_cbor_decoding_done(&ot_group, ctx);
+
+	return error;
+}
+
+const otDnsQueryConfig *otDnsClientGetDefaultConfig(otInstance *aInstance)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	bool decoded_ok;
+
+	ARG_UNUSED(aInstance);
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, 0);
+
+	nrf_rpc_cbor_cmd_rsp_no_err(&ot_group, OT_RPC_CMD_DNS_CLIENT_GET_DEFAULT_CONFIG, &ctx);
+
+	decoded_ok = ot_rpc_decode_dns_query_config(&ctx, NULL, &default_config);
+
+	nrf_rpc_cbor_decoding_done(&ot_group, &ctx);
+
+	if (!decoded_ok) {
+		nrf_rpc_err(-EBADMSG, NRF_RPC_ERR_SRC_RECV, &ot_group,
+			    OT_RPC_CMD_DNS_CLIENT_GET_DEFAULT_CONFIG, NRF_RPC_PACKET_TYPE_RSP);
+	}
+
+	return &default_config;
+}
+
+void otDnsClientSetDefaultConfig(otInstance *aInstance, const otDnsQueryConfig *aConfig)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	bool encoded;
+
+	ARG_UNUSED(aInstance);
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, OT_RPC_DNS_QUERY_CONFIG_SIZE);
+
+	encoded = ot_rpc_encode_dns_query_config(&ctx, aConfig);
+
+	if (encoded) {
+		nrf_rpc_cbor_cmd_no_err(&ot_group, OT_RPC_CMD_DNS_CLIENT_SET_DEFAULT_CONFIG, &ctx,
+					ot_rpc_decode_void, NULL);
+	}
+}
+
+static otError resolve_or_browse(enum ot_rpc_cmd_server cmd, const char *name, const char *service,
+				 void *cb, void *data, const otDnsQueryConfig *config)
+{
+	struct nrf_rpc_cbor_ctx ctx;
+	size_t to_alloc;
+	size_t name_len;
+	size_t service_len;
+
+	if (!name) {
+		return OT_ERROR_INVALID_ARGS;
+	}
+
+	name_len = strlen(name);
+
+	to_alloc = name_len + 2;
+	to_alloc += sizeof(uintptr_t) + 1;
+	to_alloc += sizeof(uint32_t) + 1;
+
+	if (service) {
+		service_len = strlen(service);
+		to_alloc += service_len + 2;
+	}
+
+	if (config) {
+		to_alloc += OT_RPC_DNS_QUERY_CONFIG_SIZE;
+	}
+
+	NRF_RPC_CBOR_ALLOC(&ot_group, ctx, to_alloc);
+
+	nrf_rpc_encode_str(&ctx, name, name_len + 1);
```

Comment:
The idea was to simply enforce having a NULL byte encoded in buffer to avoid additional code like `str[zstr.len[ = \0` in many places but I would not fight for that